Safe page sorting

pdmosses opened the pull request:
Currently, the values of `nav_order` fields are required to be uniformly all numbers or all strings. If different types of values are (accidentally) used, building fails, with a somewhat confusing error report (e.g., see issue #406).

The changes to `_includes/nav.html` eliminate the possibility of such build failures arising. They also prevent `nav_sort: case_insensitive` affecting the ordering of numerical `nav_order` parameters (where `10` came before `2`!). And they make it possible to use numbers to order some pages, and strings to order others.

The PR adds tests for various navigation features. These are excluded by default, and can be activated by commenting-out a line in `_config.yml`. (The tests inherently involve mixtures of different types of values for `nav_order` fields, so they cannot be used with previous versions of `_includes/nav.html`.)

This PR incorporates the bug fixes of PR #407 regarding `nav_exclude` and pages with numeric titles.

pdmosses commented:
@pmarsceill the Liquid code in `_includes/nav.html` is unfortunately quite complicated; suggestions for improvement are welcome! I hope the accompanying comments provide sufficient explanation, but don't hesitate to query anything that seems unmotivated.

pmarsceill reviewed (approved):
This is amazing... The tests are a great idea.

pmarsceill commented:
@pdmosses could you resolve the conflicting file from the `0.3.2` branch and I will get it merged in.

pdmosses commented:
@pmarsceill I'll take a look now, and let you know.

pmarsceill commented on _config.yml:
```diff
@@ -16,10 +16,18 @@
 description: A Jekyll theme for documentation
 baseurl: "/just-the-docs" # the subpath of your site, e.g. /blog
-url: "https://pmarsceill.github.io" # the base hostname & protocol for your site, e.g. http://example.com
+url: "https://pdmosses.github.io" # the base hostname & protocol for your site, e.g. http://example.com
```
We probably want to put this back before shipping 😅 

```suggestion
url: "https://pmarsceill.github.io" # the base hostname & protocol for your site, e.g. http://example.com
```

pdmosses replied:
Oops... indeed, it wasn't intentional 😃

pmarsceill commented on _config.yml:
```diff
@@ -16,10 +16,18 @@
+, "docs/tests/"
+]
+
+# Regression tests
```
I love this idea... A note for the future:

We could create a separate `_config.yml` file that includes the tests and then y[ou can use that instead of the default if you wanted to run the tests](https://jekyllrb.com/docs/usage/).

```
jekyll serve --config _regression_tests.yml
```

Then we wouldn't need to comment anything out, they would just live in a different file.

pdmosses replied:
Yes, that is indeed a much neater way of activating the tests locally. Next release I guess.

pdmosses commented:
@pmarsceill I've updated `_includes/nav.html` with the major changes regarding sorting and exclusion, and corrected the `url` config...

I've also updated `/docs/tests/index.md` with corrected instructions for editing `_config.yml` to activate the tests. The tests produce the intended results for me locally, but it would be good if you could check at your end. After activating the tests, the extra navigation items that appear should be consistent with all the points listed in `Tests: Navigation`. Let's switch to the neater technique you suggested in the next version.

And finally, I've reinstated the collapsible TOC at the top of the `Navigation Structure` page, to justify the reference to it in the last line of that page. If you really don't want the TOC at the top to be collapsible and revert that commit, the following sentence at the end will need adjusting or removing:
```
The result is shown at [the top of this page](#navigation-structure) (`{:toc}` can be used only once on each page).
```

pdmosses commented:
@pmarsceill the commit option on the requested change disappeared due to my independent commit, but the change request is still there... I hope it doesn't block merging. Apologies for my lack of expertise with updating PRs.
